### PR TITLE
Deploy Jenkins Instance

### PR DESCRIPTION
**Pull request by peternied**
### Description
This is built off of the hard work from @abhinavGupta16, @gaiksaya,
and @peterzhuamazon which until now was happening behind the scenes.

This change configures and deploys Jenkins running on an EC2 instance
with recommended security practices and isolation features.  For more
details consult the readme

### Check List
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: lib/compute/jenkins-main-node.ts [hunk 1]
```diff
@@ -0,0 +1,282 @@
+/**
```
Comment (abhinavGupta16) [inline lib/compute/jenkins-main-node.ts:1]: What happens to git history for the files that were moved? We should move the files in such a way that the history is preserved.

File: lib/compute/jenkins-plugins.ts [hunk 2]
```diff
@@ -0,0 +1,155 @@
+/**
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+export class JenkinsPlugins {
```
[the remark this answers is not in the document]
Comment (peterzhuamazon) [reply lib/compute/jenkins-plugins.ts:9]: I dont see this as a problem as we want to update the plugins to the latest versions to patch all vulnerabilities.
However, I also think maintaining a list of versions for stable usage is also beneficial.
You can have 1 template for latest and 1 template for stable. What do you think about this?

File: lib/compute/jenkins-main-node.ts (hunk 1 above)
Comment (peternied) [reply lib/compute/jenkins-main-node.ts:1]: We have a small amount of code and people working on this project, I don't think the cost is worthwhile.  The infra repository will maintain this history on its deleted files.

Comment (abhinavGupta16) [reply lib/compute/jenkins-main-node.ts:1]: Ohh, I didn't know we would have confidential information as a part of commit messages. Although , it is strange to have such info in commit messages, irrespective

Comment (peternied) [reply lib/compute/jenkins-main-node.ts:1]: In a private branch it doesn't matter what info we have in there :) so we can go wild.  We just need to be really really careful when pulling stuff out

File: lib/ci-stack.ts [hunk 3]
```diff
@@ -6,21 +6,58 @@
...
+} from '@aws-cdk/core';
+import { ListenerCertificate } from '@aws-cdk/aws-elasticloadbalancingv2';
+import { CIConfigStack } from './ci-config-stack';
+import { JenkinsMainNode } from './compute/jenkins-main-node';
+import { JenkinsMonitoring } from './monitoring/ci-alarms';
+import { JenkinsExternalLoadBalancer } from './network/ci-external-load-balancer';
+import { JenkinsSecurityGroups } from './security/ci-security-groups';
 
 export class CIStack extends Stack {
   // eslint-disable-next-line no-useless-constructor
```
Comment (abhinavGupta16) [inline lib/ci-stack.ts:23]: This constructor is no longer useless. We can get rid of this

File: bin/ci-stack.ts [hunk 4]
```diff
@@ -11,6 +9,10 @@
 import 'source-map-support/register';
 import * as cdk from '@aws-cdk/core';
 import { CIStack } from '../lib/ci-stack';
+import { CIConfigStack } from '../lib/ci-config-stack';
 
 const app = new cdk.App();
+
+new CIConfigStack(app, 'CI-Config-Dev', {});
```
Comment (abhinavGupta16) [inline bin/ci-stack.ts:16]: Do we need to specify the region here like we did in `Jenkins.ts`?

File: lib/ci-stack.ts [hunk 5]
```diff
@@ -6,21 +6,58 @@
...
   constructor(scope: Construct, id: string, props?: StackProps) {
     super(scope, id, props);
 
-    // Network / VPC
+    const vpc = new Vpc(this, 'JenkinsVPC');
 
-    // Security Groups
+    const useSslParameter = new CfnParameter(this, 'useSsl', {
+      description: 'If the jenkins instance should be access via SSL',
+      allowedValues: ['true', 'false'],
```
Comment (abhinavGupta16) [inline lib/ci-stack.ts:31]: Can we use `y` or `n` instead? It would be lesser syllables to type.

**Review by abhinavGupta16 — CHANGES_REQUESTED** [review]

File: lib/compute/jenkins-main-node.ts [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+      vpcSubnets: {
+        subnetType: SubnetType.PRIVATE_WITH_NAT,
+      },
+      securityGroup: props.sg,
+      init: CloudFormationInit.fromElements(...JenkinsMainNode.configElements(
+        stack.stackName,
+        stack.region,
+        props,
+      )),
+      initOptions: { ignoreFailures: true },
```
Comment (abhinavGupta16) [inline lib/compute/jenkins-main-node.ts:114]: We probably need to remove this line.

**Review by abhinavGupta16 — CHANGES_REQUESTED** [review]

File: lib/ci-stack.ts (hunk 5 above)
Comment (peternied) [reply lib/ci-stack.ts:31]: It looks like `npm run cdk deploy CI-Dev -- --parameters useSsl=false` in practice which I think is more readable.

File: lib/compute/jenkins-main-node.ts [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+      }),
+      vpc: props.vpc,
+      vpcSubnets: {
+        subnetType: SubnetType.PRIVATE_WITH_NAT,
+      },
+      securityGroup: props.sg,
+      init: CloudFormationInit.fromElements(...JenkinsMainNode.configElements(
+        stack.stackName,
+        stack.region,
+        props,
```
[the remark this answers is not in the document]
Comment (peternied) [reply lib/compute/jenkins-main-node.ts:112]: This is one of those subtle language features of typescript, super useful to make cleaner declarations

File: lib/ci-stack.ts (hunk 5 above)
Comment (abhinavGupta16) [reply lib/ci-stack.ts:31]: Would it not ask as a y/n question when deploying, something like - 
`Are you sure you want to delete: JenkinsStack (y/n)? y`?

**Pull request merged**